exporters/autoexport/registry.go

```diff
@@ -0,0 +1,151 @@
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+const (
+	otelExporterOTLPProtoEnvKey = "OTEL_EXPORTER_OTLP_PROTOCOL"
+)
+
+// registry maintains a map of exporter names to SpanExporter factories
+// func() (trace.SpanExporter) that is safe for concurrent use by multiple
+// goroutines without additional locking or coordination.
+type registry struct {
+	mu    sync.Mutex
+	names map[string]func() (trace.SpanExporter, error)
+}
+
+func newRegistry() registry {
+	return registry{
+		names: map[string]func() (trace.SpanExporter, error){
+			"":     buildOTLPExporter,
+			"otlp": buildOTLPExporter,
+		},
+	}
+}
+
+var (
+	// envRegistry is the package level registry of exporter registrations
+	// and their mapping to a SpanExporter factory func() (trace.SpanExporter, error).
+	envRegistry = newRegistry()
+
+	// errUnknownExporter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExporter = errors.New("unknown exporter")
+
+	// errInvalidOTLPProtocol is returned when an invalid protocol is used in
+	// the OTEL_EXPORTER_OTLP_PROTOCOL environment variable.
+	errInvalidOTLPProtocol = errors.New("invalid OTLP protocol - should be one of ['grpc', 'http/protobuf']")
+
+	// errDuplicateRegistration is returned when an duplicate registration is detected.
+	errDuplicateRegistration = errors.New("duplicate registration")
+)
+
+// load returns tries to find the SpanExporter factory with the key and
+// then execute the factory, returning the created SpanExporter.
+// errUnknownExporter is returned if the registration is missing and the error from
+// executing the factory if not nil.
+func (r *registry) load(key string) (trace.SpanExporter, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	factory, ok := r.names[key]
+	if !ok {
+		return nil, errUnknownExporter
+	}
+	exporter, err := factory()
+	if err != nil {
+		return nil, err
+	}
+	return exporter, nil
+}
+
+// store sets the factory for a key if is not already in the registry. errDuplicateRegistration
+// is returned if the registry already contains key.
+func (r *registry) store(key string, factory func() (trace.SpanExporter, error)) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if _, ok := r.names[key]; ok {
+		return fmt.Errorf("%w: %q", errDuplicateRegistration, key)
+	}
+	r.names[key] = factory
+	return nil
+}
+
+// drop removes key from the registry if it exists, otherwise nothing.
+func (r *registry) drop(key string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	delete(r.names, key)
+}
+
+// RegisterSpanExporter sets the SpanExporter factory to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, factory func() (trace.SpanExporter, error)) {
```

Comment:
The factory function should take a context. This will allow user values and timeouts to be passed to the factory in the future.